Data/Semiring/Tropical.hs

```diff
@@ -0,0 +1,253 @@
+import Control.Monad.Zip (MonadZip(mzipWith))
+#endif
+#if MIN_VERSION_base(4,7,0)
+import Data.Data (Data)
+#endif
+#if !MIN_VERSION_base(4,8,0)
+import Data.Foldable (Foldable)
+import Data.Traversable (Traversable)
+#endif
+#if MIN_VERSION_base(4,6,0)
+import Data.Proxy (Proxy(Proxy))
+#endif
+import Data.Semiring (Semiring(..))
+#if MIN_VERSION_base(4,7,0)
+import Data.Typeable (Typeable)
+#endif
+
+import qualified Data.Monoid as Monoid
+import qualified GHC.Natural as Natural
+
+#if !MIN_VERSION_base(4,6,0)
+-- | On older GHCs, 'Data.Proxy.Proxy' is not polykinded, so we provide our own proxy type for 'Extrema'.
+data EProxy (e :: Extrema) = EProxy
+#endif
+
+-- | A datatype to be used at the kind-level. Its only
+--   purpose is to decide the ordering for the tropical
+--   semiring in a type-safe way.
+data Extrema = Minima | Maxima
+
+-- | The 'Extremum' typeclass exists for us to match on
+--   the kind-level 'Extrema', so that we can recover
+--   which ordering to use in the `Semiring` instance
+--   for 'Tropical`.
+class Extremum (e :: Extrema) where
+  -- unfortunately this has to take a `Proxy` because
+  -- we don't have visual type applications before GHC 8.0
+#if !MIN_VERSION_base(4,6,0)
+  extremum :: EProxy e -> Extrema
+#else
+  extremum :: Proxy e -> Extrema
+#endif
+
+instance Extremum 'Minima where
+  extremum _ = Minima
+  {-# INLINE extremum #-} -- just to be safe
+
+instance Extremum 'Maxima where
+  extremum _ = Maxima
+  {-# INLINE extremum #-} -- just to be safe
+
+-- | The tropical semiring. @'Tropical' ''Minima' a@ is equivalent to the semiring
+--     \[
+--     (a \cup \{+\infty\}, \oplus, \otimes)
+--     \]
+--     where
+--     \[
+--     x \oplus y = min\{x,y\},
+--     \]
+--     \[
+--     x \otimes y = x + y.
+--     \]
+--
+--   @'Tropical' ''Maxima' a@ is equivalent to the semiring
+--     \[
+--     (a \cup \{-\infty\}, \oplus, \otimes)
+--     \]
+--     where
+--     \[
+--     x \oplus y = max\{x,y\},
+--     \]
+--     \[
+--     x \otimes y = x + y.
+--     \]
+--
+-- In literature, the 'Semiring' instance of the 'Tropical' semiring lifts
+-- the underlying semiring's additive structure. One might ask why this lifting doesn't
+-- instead witness isn't 'Monoid', since we only lift 'zero' and 'plus' - the reason is
+-- that usually the additive structure of a semiring is monotonic, i.e.
+-- @a '+' ('min' b c) == 'min' (a '+' b) (a '+' c)@, but in general this is not true.
+-- For example, lifting 'Monoid.Product' 'Natural.Natural' or 'Monoid.Product' 'Word' into 'Tropical' is lawful,
+-- but 'Monoid.Product' 'Int' is not, lacking distributivity: @(-1) '*' ('min' 0 1) '/=' 'min' ((-1) '*' 0) ((-1) '*' 1)@.
+-- So, we deviate from literature and instead
+-- witness the lifting of a 'Monoid', so the user must take care to ensure
+-- that their implementation of 'mappend' is monotonic.
+data Tropical :: Extrema -> * -> * where
+  Infinity :: Tropical e a
+  Tropical :: a -> Tropical e a
```

Comment:
What is the reason behind GADT here? Would not `data Tropical e a = Infinity | Tropical a` do the job?

Reply:
I just find GADTSyntax more pleasant to the eye. I'm unsure if there's a problem with this. Please let me know if so

Reply:
Well, UX of GADTs is much better in modern GHCs, but still not on par with vanilla ADTs. For example, with ADT you can avoid unpleasant standalone deriving below.

Reply:
By UX are you referring to deriving clauses/haddocks?

Reply:
Yes, mostly. Also there is a chance that GADTs could make life harder for tooling (IDE, linting, refactoring), present and future, but since I do not use any, I cannot enforce this claim by examples :)

It is not a big deal, of course.